### PR TITLE
Tidy endpoint code and fully initialize endpoint structure earlier

### PR DESCRIPTION
**Pull request by joestringer**
This is a series of cleanups/rearrangements of the endpoint code, primarily to simplify and co-locate similar information. Review commit-by-commit.

One change that did occur here is to extend the endpoint restore test under `daemon/state_test.go` to ensure that the entire endpoint is correctly restored, not just the MAC address.

The main goal in this PR was to consolidate the configuration of endpoints to ensure that they are properly initialized before they start regenerating, which will be required for #6618 .

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6946)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

File: pkg/endpoint/endpoint.go [hunk 1]
```diff
@@ -421,6 +420,7 @@ func NewEndpointWithState(ID uint16, state string) *Endpoint {
 		controllers:   controller.NewManager(),
 	}
 	ep.UpdateLogger(nil)
+	ep.SetDefaultOpts(option.Config.Opts)
```
Comment (aanm) [inline pkg/endpoint/endpoint.go:422]: please set defaults before updating logger, I'm not sure if the update logger users any options from the endpoint or not.

**Review by aanm — CHANGES_REQUESTED** [review]
@joestringer if you want to have more fun, the whole sort interfaces that you added can be replaced with this library function:

```golang
// Slice sorts the provided slice given the provided less function.
//
// The sort is not guaranteed to be stable. For a stable sort, use
// SliceStable.
//
// The function panics if the provided interface is not a slice.
func Slice(slice interface{}, less func(i, j int) bool) 
```

**Review by ianvernon — APPROVED** [review]

**Comment by tgraf** [comment]
Great change, let's merge it after 1.4 is out just to avoid adding unnecessary backport pain this week.

**Comment by joestringer** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Pull request merged**